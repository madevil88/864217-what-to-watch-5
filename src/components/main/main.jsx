import React from "react";
import {Link} from "react-router-dom";
import {connect} from "react-redux";
import {AuthorizationStatus} from "../../const";
import {getSelectedGenre,
  incrementFilmsCount} from "../../store/action";
import {getFilteredFilms,
  getFilms,
  getCurrentGenre,
  getFilmsCount,
  getPromoFilm,
  getAuthorizationStatus} from "../../store/selectors";
import {fetchPromoFilm} from "../../store/api-actions";
import MainProps from "./main-props";
import ListOfFilms from "../list-of-films/list-of-films";
import ListOfGenres from "../list-of-genres/list-of-genres";
import ShowMoreButton from "../show-more-button/show-more-button";
import UserBlock from "../user-block/user-block";
import MyListButton from "../my-list-button/my-list-button";
import withActiveItem from "../../hocs/with-active-item/with-active-item";

const MyListButtonWrapped = withActiveItem(MyListButton);

const Main = (props) => {
  const {films,
    getSelectedGenreAction,
    filmsCount,
    getShowMoreStatusAction,
    currentGenre,
    promoFilm,
    fetchPromoFilmAction,
    authorizationStatus} = props;

  return (
    <React.Fragment>
      <section className="movie-card">
        <div className="movie-card__bg">
          <img src={promoFilm.background_image} alt={promoFilm.name} />
        </div>
        <h1 className="visually-hidden">WTW</h1>
        <header className="page-header movie-card__head">
          <div className="logo">
            <a className="logo__link">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </a>
          </div>

          <UserBlock />
        </header>
        <div className="movie-card__wrap">
          <div className="movie-card__info">
            <div className="movie-card__poster">
              <img src={promoFilm.poster_image} alt={promoFilm.name} width="218" height="327" />
            </div>
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{promoFilm.name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{promoFilm.genre}</span>
                <span className="movie-card__year">{promoFilm.released}</span>
              </p>
              <div className="movie-card__buttons">
                <Link to={`/player/${promoFilm.id}`} className="btn btn--play movie-card__button">
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </Link>
                {(authorizationStatus === AuthorizationStatus.AUTH) &&
                  <MyListButtonWrapped
                    id={promoFilm.id}
                    initialActiveItem={promoFilm.is_favorite}
                    updateData={fetchPromoFilmAction}
                  />
                }
              </div>
            </div>
          </div>
        </div>
      </section>
      <div className="page-content">
        <section className="catalog">
          <h2 className="catalog__title visually-hidden">Catalog</h2>
          <ListOfGenres
            allFilms={films.allFilms}
            getSelectedGenre={getSelectedGenreAction}
            currentGenre={currentGenre}
          />
          <div className="catalog__movies-list">
            <ListOfFilms
              filteredFilms={films.filteredFilms}
              filmsCount={filmsCount}
            />
          </div>
          {(filmsCount < films.filteredFilms.length) && <ShowMoreButton
            getShowMoreStatus={getShowMoreStatusAction}
          />}
        </section>
        <footer className="page-footer">
          <div className="logo">
            <a className="logo__link logo__link--light">
              <span className="logo__letter logo__letter--1">W</span>
              <span className="logo__letter logo__letter--2">T</span>
              <span className="logo__letter logo__letter--3">W</span>
            </a>
          </div>
          <div className="copyright">
            <p>Â© 2019 What to watch Ltd.</p>
          </div>
        </footer>
      </div>
    </React.Fragment>
  );
};

Main.propTypes = MainProps.propTypes;

const mapStateToProps = (state) => ({
  films: {
    allFilms: getFilms(state),
    filteredFilms: getFilteredFilms(state)
  },
  currentGenre: getCurrentGenre(state),
  filmsCount: getFilmsCount(state),
  promoFilm: getPromoFilm(state),
  authorizationStatus: getAuthorizationStatus(state),
});

const mapDispatchToProps = (dispatch) => ({
  getSelectedGenreAction(selectGenre) {
    dispatch(getSelectedGenre(selectGenre));
  },
  getShowMoreStatusAction() {
    dispatch(incrementFilmsCount());
  },
  fetchPromoFilmAction() {
    dispatch(fetchPromoFilm());
  },
});

export {Main};
export default connect(mapStateToProps, mapDispatchToProps)(Main);
