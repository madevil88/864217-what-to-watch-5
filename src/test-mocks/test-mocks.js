const reviews = [{
  "id": 1,
  "user": {
    "id": 1,
    "name": `Emely`
  },
  "rating": 4,
  "comment": `This movie is just plain bad. There must be some big payola going round this awards season.`,
  "date": `2020-11-24T16:25:11.950Z`
}, {
  "id": 2,
  "user": {
    "id": 3,
    "name": `Max`
  },
  "rating": 3,
  "comment": `I personally found this movie to be boring.`,
  "date": `2020-11-27T21:02:38.606Z`
}];

const films = {
  allFilms: [{
    "name": `Bronson`,
    "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    "background_color": `#73B39A`,
    "description": `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    "rating": 3.6,
    "scores_count": 109661,
    "director": `Nicolas Winding Refn`,
    "starring": [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    "run_time": 92,
    "genre": `Action`,
    "released": 2008,
    "id": 1,
    "is_favorite": false,
    "video_link": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }, {
    "name": `Pulp Fiction`,
    "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    "background_color": `#795433`,
    "description": `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    "rating": 1.5,
    "scores_count": 1635992,
    "director": `Quentin Tarantino`,
    "starring": [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
    "run_time": 153,
    "genre": `Crime`,
    "released": 1994,
    "id": 2,
    "is_favorite": true,
    "video_link": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }],
  filteredFilms: [{
    "name": `Bronson`,
    "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    "background_color": `#73B39A`,
    "description": `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    "rating": 3.6,
    "scores_count": 109661,
    "director": `Nicolas Winding Refn`,
    "starring": [`Tom Hardy`, `Kelly Adams`, `Luing Andrews`],
    "run_time": 92,
    "genre": `Action`,
    "released": 2008,
    "id": 1,
    "is_favorite": false,
    "video_link": `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }, {
    "name": `Pulp Fiction`,
    "poster_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    "background_image": `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    "background_color": `#795433`,
    "description": `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    "rating": 1.5,
    "scores_count": 1635992,
    "director": `Quentin Tarantino`,
    "starring": [`John Travolta`, `Uma Thurman`, `Samuel L. Jackson`],
    "run_time": 153,
    "genre": `Crime`,
    "released": 1994,
    "id": 2,
    "is_favorite": true,
    "video_link": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    "preview_video_link": `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }]
};

export {
  reviews,
  films
};
