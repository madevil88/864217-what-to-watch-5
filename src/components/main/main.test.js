import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import {Main} from "./main";
import {films} from "../../test-mocks/test-mocks";

const renderer = new ShallowRenderer();
const currentGenre = films.allFilms[1].genre;

it(`Should Main render correctly`, () => {
  renderer.render(
      <Main
        films={films}
        getSelectedGenreAction={() => {}}
        filmsCount={8}
        getShowMoreStatusAction={() => {}}
        currentGenre={currentGenre}
        promoFilm={films.allFilms[0]}
        fetchPromoFilmAction={() => {}}
      />
  );

  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});
