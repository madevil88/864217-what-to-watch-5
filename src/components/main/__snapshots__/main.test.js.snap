// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Main render correctly 1`] = `
<React.Fragment>
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Bronson"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <Memo(Connect(UserBlock)) />
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="Bronson"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Bronson
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Action
            </span>
            <span
              className="movie-card__year"
            >
              2008
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <Link
              className="btn btn--play movie-card__button"
              to="/player/1"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </Link>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ListOfGenres
        allFilms={
          Array [
            Object {
              "background_color": "#73B39A",
              "background_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg",
              "description": "A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.",
              "director": "Nicolas Winding Refn",
              "genre": "Action",
              "id": 1,
              "is_favorite": false,
              "name": "Bronson",
              "poster_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg",
              "preview_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg",
              "preview_video_link": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 3.6,
              "released": 2008,
              "run_time": 92,
              "scores_count": 109661,
              "starring": Array [
                "Tom Hardy",
                "Kelly Adams",
                "Luing Andrews",
              ],
              "video_link": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
            },
            Object {
              "background_color": "#795433",
              "background_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg",
              "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
              "director": "Quentin Tarantino",
              "genre": "Crime",
              "id": 2,
              "is_favorite": true,
              "name": "Pulp Fiction",
              "poster_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg",
              "preview_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg",
              "preview_video_link": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 1.5,
              "released": 1994,
              "run_time": 153,
              "scores_count": 1635992,
              "starring": Array [
                "John Travolta",
                "Uma Thurman",
                "Samuel L. Jackson",
              ],
              "video_link": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
            },
          ]
        }
        currentGenre="Crime"
        getSelectedGenre={[Function]}
      />
      <div
        className="catalog__movies-list"
      >
        <ListOfFilms
          filmsCount={8}
          filteredFilms={
            Array [
              Object {
                "background_color": "#73B39A",
                "background_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg",
                "description": "A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.",
                "director": "Nicolas Winding Refn",
                "genre": "Action",
                "id": 1,
                "is_favorite": false,
                "name": "Bronson",
                "poster_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg",
                "preview_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg",
                "preview_video_link": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 3.6,
                "released": 2008,
                "run_time": 92,
                "scores_count": 109661,
                "starring": Array [
                  "Tom Hardy",
                  "Kelly Adams",
                  "Luing Andrews",
                ],
                "video_link": "http://media.xiph.org/mango/tears_of_steel_1080p.webm",
              },
              Object {
                "background_color": "#795433",
                "background_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg",
                "description": "The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
                "director": "Quentin Tarantino",
                "genre": "Crime",
                "id": 2,
                "is_favorite": true,
                "name": "Pulp Fiction",
                "poster_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg",
                "preview_image": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg",
                "preview_video_link": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
                "rating": 1.5,
                "released": 1994,
                "run_time": 153,
                "scores_count": 1635992,
                "starring": Array [
                  "John Travolta",
                  "Uma Thurman",
                  "Samuel L. Jackson",
                ],
                "video_link": "http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4",
              },
            ]
          }
        />
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</React.Fragment>
`;
